### PR TITLE
SQL Lesson

### PR DESCRIPTION
**Pull request by adambray**

### CODE REVIEW

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
+# Databases
+
+## Objectives
+
+### Concepts
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: or, learning objectives

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+**Consistency** - Databases can enforce rules regarding consitency of data,
+especially when handling simultaneous requests to update information
+
+**Scalability** - Databases can handle a lot more request per second, and many
+DBs have ways to scale to massive loads by replicating / syncing information
+across multiple DBs
+
+**Querying** - DBs make it easy to search, sort, filter, and combine related
+data using a Query Language.
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: maybe mention the ACID acronym here? https://en.wikipedia.org/wiki/ACID

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+## What's a relational database?
+
+The most popular type of database is a **relational** database.
+
+Relational databases:
+
+- Store data in tables, which have columns and rows, much like a spreadsheet
+  - Tables might be `songs` and `artists`
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: nice foreshadowing

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+- Communicate via SQL (Structured Query Language)
+- Can relate data between tables (hence the name *relational* database)
+  - e.g. we can relate rows in the `songs` table to rows in the `artists` table
+- We often use a `key`, which is a field that is unique for each row in a table.
+  - `id` is a common key, starting at 1.
+
+There are lots of relational databases, such as PostgreSQL, MySQL, SQLite, etc.
+They are all pretty similar (use SQL), but do have different features.
+
+PostgreSQL is the most popular open-source database for web development.
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: got any stats or evidence of this?

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+- List common data types used in SQL and the related field constraints
+- Describe what a SQL injection attack is
+
+### Mechanics
+
+- Create a new PostgreSQL database
+- Set up a PostgreSQL database schema with a saved SQL file
+- Seed a PostgreSQL database with a saved SQL file
+- Execute basic SQL commands to execute CRUD actions in a database
+
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: add opening framing section. what are we talking about today? why now? how does this relate to what came before / what's to come?

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+  "Not Only SQL". Store data without using tables. More like 'big buckets'.
+  We'll cover these later in the course.
+
+## Exploring Postgres
+
+Launch `Postgres.app` if you don't see the elephant in your menu bar.
+
+### psql commands
+
+We'll use `psql` as our primary means of interacting with our databases.
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: eventually, we'll use ruby

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+### Defining a Schema
+
+We need to define a schema for our database. It can change later on if we need
+to add / remove tables or columns, but we'll start with something simple.
+
+Instead of typing this into psql, it's easier to save it into a file, and 'run'
+that file.
+
+We'll be using this as an example today:
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: mention something about the file ending in `.sql` and not `.rb` - very common error

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+
+[Library SQL Exercise](https://github.com/ga-dc/library_sql)
+
+#### Creating our database
+
+```bash
+$ createdb library
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: point out this is an alternate method from before

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+  puts author["name"] + " " + author["birth_year"]
+end
+```
+
+## Security!
+
+- What security holes do you see here?
+  - Let's say I want to update the database with something a user writes into their computer...
+  - They could pretty easily make my code execute a DROP TABLE or something.
+  - SQL injection
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: link to external resource for further reading?

File: 04-ruby-mvc-sinatra/databases/sql_basics.md [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+  - [Obligatory visual aid](images/xkcd.png)
+  - [Other obligatory visual aid](images/car.jpg)
+
+```ruby
+connection.prepare('insert_student_statement', 'INSERT INTO authors (name, nationality, birth_year) VALUES ($1, $2, $3)')
+connection.exec_prepared('insert_student_statement', [ 'Jesse Shawl', 'Mars', 2001])
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: wtf? ha

File: 04-ruby-mvc-sinatra/databases/sql_relationships.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+
+## Joins
+
+- Inner Join
+  - Combines records that match both sides
+- Left/Right Join
+  - Combines records of one side with matching record from other side, and fills in null for remaining
+- Outer join
+  - Combines all records of both sides with matching record from other side, and fills in null for remaining.
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/databases/sql_relationships.md]: link to http://blog.codinghorror.com/a-visual-explanation-of-sql-joins/

File: 04-ruby-mvc-sinatra/domain_modeling/readme.md [hunk 12]
```diff
@@ -51,32 +54,14 @@ this data.  We use it to decide what needs to be persisted.
...
+As a class, let's come up with an example ERD for an application that manages
+a library.
 
-* Books
-	* title
-	* author
-	* year
-	* genre
-	* pages
+### Exercise: Tunr ERD
```
Comment (jshawl) [inline 04-ruby-mvc-sinatra/domain_modeling/readme.md]: maybe show completed application here?

File: 04-ruby-mvc-sinatra/databases/sql_basics.md (hunk 2 above)
Comment (adambray) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: :+1:

File: 04-ruby-mvc-sinatra/databases/sql_basics.md (hunk 10 above)
Comment (adambray) [inline 04-ruby-mvc-sinatra/databases/sql_basics.md]: want me to change this? :)

**Pull request merged**